### ISSUE TITLE
Solve collisions between ref names and test-tree methods

### ISSUE THREAD

**Issue by oliverwoodings** [QubitProducts/react-test-tree#47]
We have an inherent flaw in test-tree's node API design. @jhollingworth described it well: there are two separate domains that are able to influence the API of a node. The first is test-tree's node methods themselves, and the second is the collection of refs of that node. This means it is possible to have collisions between ref names and method names. For example, some components in React Bootstrap have refs called `value` which collide with test-tree's `value` getter.

@jhollingworth has proposed that refs no longer be applied to the node's top-level API, and instead are accessed through `get` and `getIn` methods. For example:

_Current:_
```
var tree = testTree(<MyComponent />);
tree.foo.bar.click();
```

_Future:_
```
var tree = testTree(<MyComponent />);
tree.get("foo").get("bar").click();
// or
tree.getIn(["foo", "bar"]).click();
// or
tree.getIn("foo", "bar").click();
```

This would obviously have to be released as a major breaking change, so I would want to add a deprecation notice to the next minor version (when we add support for react 0.14) alongside adding in the get and getIn methods, and then fully deprecate refs in a major release after that.

@giuse88 @wbecker you guys have been bitten by this issue of colliding ref names before. What do you think of this proposal?

**Comment by wbecker** [comment]
The beauty of react test tree is that its very terse to get at things and simulate them. Adding this in would remove that and would suck - especially if its just for a few edge cases that don't normally hit you. 

Could you instead have some way of specifying aliases - and detecting when a collision happens without an alias so that the user knows he ought to add one in?

**Comment by oliverwoodings** [comment]
I guess it could be possible to specify aliases in the same way you can stub things out. E.g:

```
testTree(<Blah />, {
  alias: {
    foo: {
      bar: "baz"
    }
  }
});
```

If we choose this option I would want to move static property access to methods. E.g. instead of `node.value` you do `node.getValue`. Thoughts on that?

**Comment by wbecker** [comment]
Static as in the special helpers that react-test-tree provides? I guess
that's not so bad.

**Comment by oliverwoodings** [comment]
Yeah - the majority of collisions come from react-test-tree's `element` and `value` properties

**Comment by jhollingworth** [comment]
That API is horrible

![](https://media.giphy.com/media/rOJvPDsG3StAk/giphy.gif)

**Comment by oliverwoodings** [comment]
@jhollingworth which bit?

**Comment by oliverwoodings** [comment]
@jhollingworth just following up on this again, going to try and get test tree updated for 0.14 and also try and get this issue closed. Which bit of that proposed API didn't you like?

**Comment by Olical** [comment]
I personally really like the `get` / `getIn` style since it's consistent with the ImmutableJS API. The alternative is to just namespace the properties to `refs` or something.

**Comment by oliverwoodings** [comment]
Yeah I've considered namespacing them to `refs`, however it becomes a bit confusing. You can do this:

`tree.refs.bar.refs.foo.element.refs`

Which would make you think that the last `refs` is the same as the others, except it isn't since it is the raw refs from the element.

When written with the `get/getIn` style you end up with this:

`tree.getIn("bar", "foo").element.refs`

Which I am pretty happy with now.

**Comment by Olical** [comment]
Sounds good to me. Worth keeping it consistent with the ImmutableJS API and passing an array to `{get,set}In` as opposed to using arguments? Allows for extension of the API later.

I really like it too since a path is always represented as an array in every instance making it easier to spot and work with. It encourages you to define paths ahead of time then too.

**Comment by oliverwoodings** [comment]
Yeah good point, I'll restrict it to an array instead of separate args.

**Comment by oliverwoodings** [comment]
Closed by #31
